New validator_info and manifest RPC methods

movitto opened the pull request:
This pull request addresses issue #3099 by adding two new RPC methods to allow the inspection of validator state.

- validator_info will return information pertaining to the current validators keys, manifest sequence, and domain

- manifest looks up manifest information for the specified key (either master or ephemeral)

Utility methods have been added to other modules to support this functionality

nbougalis commented on src/ripple/app/misc/Manifest.h:
```diff
@@ -257,6 +257,32 @@ class ManifestCache
 
+    /** Returns master key's current manifest sequence.
+
+        @return sequence corresponding to Master public key
```
Micro-nit: no need for multiple **`@return`** statements:

```
@return sequence corresponding to master public key if
        configured or 0 otherwise.
```

I also appreciate the thread-safety notation, but I'm not sure that it adds a lot. Of course, I'm biased (since I understand the code) and perhaps others will find it useful.

movitto replied:
Merged the @return statements into one. As far as thread safety, was following the style demonstrated in the other accessor methods but removed in the followup patch.

nbougalis commented on src/ripple/app/misc/Manifest.h:
```diff
@@ -257,6 +257,32 @@ class ManifestCache
+
+    /** Returns master key's current domain
+
+        @return domain corresponding to Master public key
```
Similar comment as above.

movitto replied:
Done

nbougalis commented on src/ripple/app/misc/Manifest.h:
```diff
@@ -257,6 +257,32 @@ class ManifestCache
+    std::uint32_t
+    getSequence (PublicKey const& pk) const;
+
+    /** Returns master key's current domain
```
Not to get into a grammar war, but this possessive is likely to irritate someone at some point. It also reads slightly weird, imo. Consider:

`/** Return the domain claimed by a given public key

movitto replied:
Done

nbougalis commented on src/ripple/app/misc/impl/Manifest.cpp:
```diff
@@ -309,6 +309,30 @@ ManifestCache::getMasterKey (PublicKey const& pk) const
+    std::lock_guard lock{read_mutex_};
+    auto const iter = map_.find (pk);
+
+    if (iter != map_.end () && !iter->second.revoked ())
```
I'm not sure if returning `0` is appropriate for a revoked manifest.

movitto replied:
Changed this to boost::optional and boost::none as suggested below

nbougalis commented on src/ripple/net/impl/RPCCall.cpp:
```diff
@@ -144,6 +145,21 @@ class RPCParser
         }
     }
 
+    static Json::Value validPublicKey (std::string const& strPk)
```
This function should return a `bool` and not a `Json::Value`.

movitto replied:
Done

nbougalis commented on src/ripple/net/impl/RPCCall.cpp:
```diff
@@ -144,6 +145,21 @@ class RPCParser
+        if (!publicKeyType(makeSlice(*pkHex)))
+            return false;
+
+        return true;
```
An alternative here is this:

```
if (auto k = strUnHex(strPk))
    return static_cast<bool>(publicKeyType(makeSlice(*k)));
return false;
```

Note that this rewrite returns `true` from the branch that _positively_ identifies a key as being of a given type, instead of by default. I think that's a net positive. Apart from that, I don't know if it's better; it is _shorter_, but that isn't always a positive.

movitto replied:
@nbougalis do you mind if we leave this as is for this patch? The reason I suggest this is that this method is simply extracted verbatim from the **parseFetchInfo** method below. This way we can just focus on extracting the common logic here, and refactoring / cleaning this up in a separate PR.

nbougalis replied:
Not at all. Whether review comments should be "resolved" is typically up to the PR author. Obviously, exceptions for bugs/problems exist, but this wasn't that. I'm fine with leaving this as is.

nbougalis commented on src/ripple/net/impl/RPCCall.cpp:
```diff
@@ -202,6 +218,25 @@ class RPCParser
+          if (!validPublicKey (strPk))
+            return rpcError (rpcPUBLIC_MALFORMED);
+
+          jvRequest[jss::public_key] = strPk;
```
If `validPublicKey` returned `boost::optional<PublicKey>` then we could return the public key in "canonical" format (whether base58 or hex) instead of whichever format the user provided it for us.

movitto replied:
Done

nbougalis commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -0,0 +1,46 @@
+    Json::Value ret;
+
+    auto pk = context.app.validatorKeys().publicKey;
+    ret[jss::master_key]        = toBase58(TokenType::NodePublic, pk);
```
Similarly, why not:

```
ret[jss::master_key] = toBase58(TokenType::NodePublic,
    context.app.getValidationPublicKey());
```

Also hate to complain about whitespace, but (here and elsewhere) I'd put one space before and one space after the `=` and avoid attempting to "align" the code manually.

movitto replied:
Done

nbougalis commented on src/ripple/rpc/handlers/Manifest.cpp:
```diff
@@ -0,0 +1,60 @@
+    auto ek = context.app.validatorManifests().getSigningKey(mk);
+
+    // if ephemeral key not found, we don't have specified manifest
+    if(ek == mk)
```
This is a particularly confusing way of determining that we don't have a given manifest; not your fault, the API for this kind of thing is just awkward and it's something we ought to look at improving.

It also results in returning an empty JSON object and that _is_ something we can address here and now: we should, at the very least, return something helpful, e.g. the key that we searched for.

Consider if the returned JSON was:

```
{
    "requested": "nHDpbBXLzgktPjgcEEpb1Q9YgNDnZBPdHRXHza8PNLbUpGFxBZxV",
    /** These appear only if the manifest was found */
    "manifest" : "the actual manifest in hex",
    "details" : {
        "master_key" : "nHDpbBXLzgktPjgcEEpb1Q9YgNDnZBPdHRXHza8PNLbUpGFxBZxV",
        "ephemeral_key" : "n9LdwcMvLtsgpVcNeQJDCTQWbd46AY1i71pQPiJPsjnjBYuPWMu7",
        "sequence" : 1,
        "domain" : "area51.number8.us"
    }
}
```

movitto replied:
Updated the return JSON to match the suggested format and to always return the 'requested' key even in the case of the failed lookup

nbougalis commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -0,0 +1,46 @@
+Json::Value doValidatorInfo (RPC::Context& context)
+{
+    // return error if not configured as validator
+    if(context.app.validatorKeys().publicKey.empty())
```
Why not: `if(context.app.getValidationPublicKey().empty())`?

movitto replied:
Done

nbougalis commented on src/ripple/app/main/Application.h:
```diff
@@ -137,6 +138,7 @@ class Application : public beast::PropertyStream::Source
     virtual TxQ&                        getTxQ() = 0;
     virtual ValidatorList&              validators () = 0;
     virtual ValidatorSite&              validatorSites () = 0;
+    virtual ValidatorKeys const&        validatorKeys () = 0;
```
This seems unnecessary to me for this patch. See why further below.

movitto replied:
Done

nbougalis commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -0,0 +1,46 @@
+    ret[jss::master_key]        = toBase58(TokenType::NodePublic, pk);
+
+    auto ek = context.app.validatorManifests().getSigningKey(pk);
+    ret[jss::ephemeral_key]     = toBase58(TokenType::NodePublic, ek);
```
You should detect and account for the use case where a validator is configured with a fixed key and not a manifest. If that happens, this field and those that follow are not relevant and should not be included.

movitto replied:
I **believe** I addressed this by checking if the ephemeral key does exists, and if not using this to imply the manifest is not present. Please correct if this is the wrong assumption

nbougalis commented on src/ripple/app/misc/impl/Manifest.cpp:
```diff
@@ -309,6 +309,30 @@ ManifestCache::getMasterKey (PublicKey const& pk) const
+    return 0;
+}
+
+std::string
```
This and `ManifestCache::getSequence` should probably wrap their return in a `boost::optional`, making it possible to distinguish a "I can't find the key you asked for" from "the key you asked for doesn't have the attribute you request".

movitto replied:
Done

seelabs commented on src/ripple/net/impl/RPCCall.cpp:
```diff
@@ -114,6 +114,7 @@ class RPCParser
         return true;
     }
 
+
```
Extra space snuck in here.

seelabs commented on src/ripple/net/impl/RPCCall.cpp:
```diff
@@ -202,6 +218,25 @@ class RPCParser
 
+    Json::Value parseManifest (Json::Value const& jvParams)
+    {
+        if (1 == jvParams.size ())
```
Style: We don't use yoda conditions in rippled. I understand the reasoning, but I'd like to stay consistent with the style of the rest of rippled.

seelabs commented on src/ripple/net/impl/RPCCall.cpp:
```diff
@@ -202,6 +218,25 @@ class RPCParser
+    {
+        if (1 == jvParams.size ())
+        {
+          Json::Value     jvRequest (Json::objectValue);
```
Two style nits:
1) should indent four spaces
2) should be only one space after the type

seelabs commented on src/ripple/rpc/handlers/Manifest.cpp:
```diff
@@ -0,0 +1,75 @@
+
+    auto manifest = context.app.validatorManifests().getManifest(mk);
+    if(manifest)
+      ret[jss::manifest] = base64_encode((std::uint8_t const*)manifest->c_str(),
```
Two style nits:
1) should be indented four spaces
2) we usually use `static_cast` over c-style casts

seelabs replied:
@movitto (responding to your comment here https://github.com/ripple/rippled/pull/3197#issuecomment-572272128)

`static_cast` won't work - we need a `reinterpret_cast` so something like:
```
        ret[jss::manifest] = base64_encode(
            reinterpret_cast<std::uint8_t const*>(manifest->data()),
            manifest->length());
```

*However*, there's an overload of `base64_encode` that takes a `std::string`; I'd use that. So we can change this to:
```
        ret[jss::manifest] = base64_encode(*manifest);
```

seelabs commented on src/ripple/rpc/handlers/Manifest.cpp:
```diff
@@ -0,0 +1,75 @@
+      return ret;
+
+    auto manifest = context.app.validatorManifests().getManifest(mk);
+    if(manifest)
```
If a variable is only used in the if statement, I prefer to put the declaration in the if statment. `if (auto manifest = context....)`. No change needed if you really hate that style, but I do prefer to keep variable scopes as narrow as we can (in general). This is done in several places in this file (and ValidatorInfo.cpp). I'll briefly tag the others as well.

seelabs commented on src/ripple/rpc/handlers/Manifest.cpp:
```diff
@@ -0,0 +1,75 @@
+    details[jss::ephemeral_key] = toBase58(TokenType::NodePublic, ek);
+
+    auto seq = context.app.validatorManifests().getSequence(mk);
+    if(seq)
```
`seq`'s scope could be narrowed to the `if` statement.

Also, four spaces indent.

seelabs commented on src/ripple/rpc/handlers/Manifest.cpp:
```diff
@@ -0,0 +1,75 @@
+      details[jss::seq] = *seq;
+
+    auto domain = context.app.validatorManifests().getDomain(mk);
+    if(domain)
```
`domain` scope could be narrowed to the if statement

seelabs commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -0,0 +1,61 @@
+    ret[jss::ephemeral_key] = toBase58(TokenType::NodePublic, ek);
+
+    auto manifest = context.app.validatorManifests().getManifest(mk);
+    if(manifest)
```
`manifest` scope could be narrowed to the `if` statement

Also, indent four spaces.

seelabs commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -0,0 +1,61 @@
+                                                             manifest->length());
+
+    auto seq = context.app.validatorManifests().getSequence(mk);
+    if(seq)
```
`seq` scope could be narrowed to the if statement

Also one space after `if` and condition and indent four spaces

seelabs commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -0,0 +1,61 @@
+      ret[jss::seq] = *seq;
+
+    auto domain = context.app.validatorManifests().getDomain(mk);
+    if(domain)
```
`domain` scope could be narrowed to the if statement.

Also, four spaces indent

seelabs commented:
I also noticed there are no unit tests for this. It would be good to have some simple tests for this as well.

seelabs commented:
Non-unity build is broken. I think if both ValidatorInfo and Manifest have the following includes it should fix it:
```
 
+#include <ripple/app/main/Application.h>
 #include <ripple/basics/base64.h>
+#include <ripple/json/json_value.h>
+#include <ripple/protocol/ErrorCodes.h>
+#include <ripple/protocol/jss.h>
+#include <ripple/rpc/Context.h>
```

movitto commented:
@seelabs thanks for the review. Resolved the style issues and added missing includes for the unity build.

As far as static_cast, when making the change results in the following error:

```
/home/mmorsi/Workspace/rippled/src/ripple/rpc/handlers/Manifest.cpp:55:94: error: invalid static_cast from type ‘const char*’ to type ‘const uint8_t*’ {aka ‘const unsigned char*’}
   55 |         ret[jss::manifest] = base64_encode(static_cast<std::uint8_t const*>(manifest->c_str()),
```
Could you advice a fix or should this be left as is?

Will look into adding some tests in the near future.

jwbusch commented:
> /home/mmorsi/Workspace/rippled/src/ripple/rpc/handlers/Manifest.cpp:55:94: error: invalid static_cast from type ‘const char*’ to type ‘const uint8_t*’ {aka ‘const unsigned char*’}
>    55 |         ret[jss::manifest] = base64_encode(static_cast<std::uint8_t const*>(manifest->c_str()),
> ```
> 
> Could you advice a fix or should this be left as is?
> 
> Will look into adding some tests in the near future.

` ret[jss::manifest] = base64_encode(*manifest);`

Should work as @seelabs mentioned.

movitto commented:
@seelabs @jwbusch cool thanks for the feedback. Called the simpler base64_encode method, removed the cast, and added some tests. Let me know if there is anything else!

jwbusch commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -0,0 +1,63 @@
+
+    Json::Value ret;
+
+    auto pk = context.app.getValidationPublicKey();
```
nit: Would like to see more `const` used if possible. i.e. `const auto pk = ...`

Makes it clear that the variable won't be modified later (which is the case for a lot of these).

jwbusch commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -0,0 +1,63 @@
+
+    // lookup ephemeral key
+    auto ek = context.app.validatorManifests().getSigningKey(mk);
+    ret[jss::ephemeral_key] = toBase58(TokenType::NodePublic, ek);
```
this one could be scoped like the others down below

```
if (const auto ek = ...)
   ret[...] = ...;
```

movitto commented:
@jwbusch pushed a patch constantizing the fields. As far as the scope ephimeral_key, that method (getSigningKey, already existing before this PR) works a bit different than the new ones in the sense that if the key is not found, the input param (the master key) is simply returned. Since this is not a boost::optional we can't check it's 'truthyness' like the other fields.

seelabs commented on Builds/CMake/RippledCore.cmake:
```diff
@@ -979,6 +981,7 @@ else ()
     src/test/rpc/LedgerData_test.cpp
     src/test/rpc/LedgerRPC_test.cpp
     src/test/rpc/LedgerRequestRPC_test.cpp
+    src/test/rpc/Manifest_test.cpp
```
I think this needs to be `ManifestRPC_test.cpp`

movitto replied:
@seelabs doh! :man_facepalming: fixed

seelabs commented:
I think this is ready to go, we just need to fix the non-unity build issues. I think we just need to patch two files. I think this patch should fix things: https://github.com/seelabs/rippled/commit/b47e3742cdff7062d56add2ac3d2a97d541f82a5

movitto commented:
> I think this is ready to go, we just need to fix the non-unity build issues. I think we just need to patch two files. I think this patch should fix things: [seelabs@b47e374](https://github.com/seelabs/rippled/commit/b47e3742cdff7062d56add2ac3d2a97d541f82a5)

@seelabs cherry picked onto this branch, thanks!

seelabs commented:
@movitto FYI: Rebasing should be pretty straight-forward. One change to look out for is the parameters for `doValidatorInfo` and `doManifest` need to be changed from `RPC::Context` to `RPC::JsonContext`.

nbougalis commented:
@movitto: I can handle the rebase and squashing down if you want, and would be happy to jump on a call with you to walk you through the commands and explain what I'm doing so you know for the next time.

movitto commented:
@nbougalis sure that works. With the XRP meetup tomorrow won't have time to do the call until later on in the week, but perhaps we can schedule it for then (lets coordinate on that one via email)

movitto commented:
@seelabs @nbougalis rebased and pushed a followup patch w/ the 'Context' -> 'JsonContext' substitution. Using @mellery451's protobuf fix from #3244 (not included in this PR) I was able to build and test locally.

seelabs commented:
@movitto Things looks good; I just added a "passed" label. Thanks for (yet another) contribution!
